runtime/compiler/aarch64/codegen/CallSnippet.cpp

```diff
@@ -0,0 +1,590 @@
+ * is available at https://www.apache.org/licenses/LICENSE-2.0.
+ *
+ * This Source Code may also be made available under the following
+ * Secondary Licenses when the conditions for such availability set
+ * forth in the Eclipse Public License, v. 2.0 are satisfied: GNU
+ * General Public License, version 2 with the GNU Classpath
+ * Exception [1] and GNU General Public License, version 2 with the
+ * OpenJDK Assembly Exception [2].
+ *
+ * [1] https://www.gnu.org/software/classpath/license.html
+ * [2] http://openjdk.java.net/legal/assembly-exception.html
+ *
+ * SPDX-License-Identifier: EPL-2.0 OR Apache-2.0 OR GPL-2.0 WITH Classpath-exception-2.0 OR LicenseRef-GPL-2.0 WITH Assembly-exception
+ *******************************************************************************/
+
+#include "codegen/CallSnippet.hpp"
+#include "codegen/ARM64AOTRelocation.hpp"
+#include "codegen/CodeGenerator.hpp"
+#include "codegen/InstOpCode.hpp"
+#include "codegen/Linkage.hpp"
+#include "codegen/Machine.hpp"
+#include "codegen/Register.hpp"
+#include "env/CompilerEnv.hpp"
+#include "env/J2IThunk.hpp"
+#include "env/VMJ9.h"
+#include "il/Node.hpp"
+#include "il/Node_inlines.hpp"
+#include "il/symbol/LabelSymbol.hpp"
+#include "runtime/CodeCacheManager.hpp"
+
+static uint8_t *storeArgumentItem(TR::InstOpCode::Mnemonic op, uint8_t *buffer, TR::RealRegister *reg, int32_t offset, TR::CodeGenerator *cg)
+   {
+   TR::RealRegister *stackPtr = cg->getLinkage()->cg()->getStackPointerRegister();
+   TR::InstOpCode opCode(op);
+   uint32_t enc = (uint32_t)opCode.getOpCodeBinaryEncoding();
+   TR_ASSERT((enc & 0x3b200000) == 0x39000000, "Instruction not supported in storeArgumentItem()");
+
+   uint32_t size = (enc >> 30) & 3; /* b=0, h=1, w=2, x=3 */
+   uint32_t shifted = offset >> size;
+   uint32_t *wcursor = (uint32_t *)buffer;
+
+   opCode.copyBinaryToBuffer(buffer);
+   reg->setRegisterFieldRT(wcursor);
+   stackPtr->setRegisterFieldRN(wcursor);
+   *wcursor |= (shifted & 0xfff) << 10; /* imm12 */
+
+   return buffer+4;
+   }
+
+static uint8_t *flushArgumentsToStack(uint8_t *buffer, TR::Node *callNode, int32_t argSize, TR::CodeGenerator *cg)
+   {
+   uint32_t intArgNum=0, floatArgNum=0, offset;
+   TR::Machine *machine = cg->machine();
+   const TR::ARM64LinkageProperties &linkage = cg->getLinkage(callNode->getSymbol()->castToMethodSymbol()->getLinkageConvention())->getProperties();
+   int32_t argStart = callNode->getFirstArgumentIndex();
+
+   if (linkage.getRightToLeft())
+      offset = linkage.getOffsetToFirstParm();
+   else
+      offset = argSize+linkage.getOffsetToFirstParm();
+
+   for (int32_t i=argStart; i<callNode->getNumChildren();i++)
+      {
+      TR::Node *child = callNode->getChild(i);
+      switch (child->getDataType())
+         {
+         case TR::Int8:
+         case TR::Int16:
+         case TR::Int32:
+            if (!linkage.getRightToLeft())
+               offset -= 4;
```

Comment:
Rather than the constant `4`, use `TR::Compiler->om.sizeofReferenceAddress()` instead.  This applies to all other uses below.

`4` is actually incorrect because on 64-bit platforms the slot size is the size of a reference (i.e., 8).

Reply:
Changed the code using `TR::Compiler->om.sizeofReferenceAddress()` based on the `p` version.